ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java

```diff
@@ -0,0 +1,196 @@
+  }
+
+  private long getNumDirectTotalDeallocations() {
+    return numDirectTotalDeallocations;
+  }
+
+  private long getNumDirectTotalActiveAllocations() {
+    return numDirectTotalActiveAllocations;
+  }
+
+  private void register() {
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectArenas"),
+        (Gauge<Integer>) () -> getNumDirectArenas());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberHeapArenas"),
+        (Gauge<Integer>) () -> getNumHeapArenas());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberThreadLocalCaches"),
+        (Gauge<Integer>) () -> getNumThreadLocalCaches());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "UsedHeapMemory"),
+        (Gauge<Long>) () -> getUsedHeapMemory());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "UsedDirectMemory"),
+        (Gauge<Long>) () -> getUsedDirectMemory());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberHeapTotalAlloocations"),
+        (Gauge<Long>) () -> getNumHeapTotalAllocations());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberHeapTotalDealloocations"),
+        (Gauge<Long>) () -> getNumHeapTotalDeallocations());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberHeapTotalActiveAlloocations"),
+        (Gauge<Long>) () -> getNumHeapTotalActiveAllocations());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectTotalAlloocations"),
+        (Gauge<Long>) () -> getNumDirectTotalAllocations());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectTotalDealloocations"),
+        (Gauge<Long>) () -> getNumDirectTotalDeallocations());
+    registry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectTotalActiveAlloocations"),
+        (Gauge<Long>) () -> getNumDirectTotalActiveAllocations());
+  }
+
+  /**
+   * Start to collect metrics.
+   */
+  public void start() {
+    if (started.compareAndSet(false, true)) {
+      register();
+      scheduler = Utils.newScheduler(1, false);
+      scheduler.scheduleAtFixedRate(new NettyMetricCollector(), 0, config.nettyMetricsRefreshIntervalSeconds,
+          TimeUnit.SECONDS);
+      logger.trace("Schedule netty metric collector");
+    }
+  }
+
+  /**
+   * Stop collecting metrics. This method would stop this singleton effectively. Do not reuse this single after
+   * calling this method.
+   */
+  public void stop() {
+    if (started.compareAndSet(true, false)) {
+      if (scheduler != null) {
+        Utils.shutDownExecutorService(scheduler, config.nettyMetricsStopWaitTimeoutSeconds, TimeUnit.SECONDS);
+        logger.trace("De-scheduled for collecting netty metrics");
+      }
+    }
+  }
+
+  private class NettyMetricCollector implements Runnable {
+
+    @Override
+    public void run() {
+      PooledByteBufAllocatorMetric metric = PooledByteBufAllocator.DEFAULT.metric();
```

Comment:
please ensure when the method is first called, the metric is not null. If the config is off (netty bytebuf is not enabled) should we start NettyMetrics?

Reply:
PooledByteBuf would create metric even if we don't use it. As long as we import this library, the metric will be created. I will not start NettyMetric if we don't enable it in Network layer.